### ISSUE TITLE
ttnn.clamp (tensor bounds) accepts output_tensor= and never writes it: the caller's buffer is left stale while the scalar-bounds overload honours the same argument

### ISSUE THREAD

**Issue by singhharsh1708** [tenstorrent/tt-metal#55334]
**Component / Area:** ttnn eltwise unary (composite)

**Issue Type:** Bad Outputs

## Observed

`ttnn.clamp` accepts `output_tensor=` on both of its overloads. The scalar-bound overload writes into the caller's buffer and returns it. The tensor-bound overload allocates a fresh tensor, returns that, and never touches the buffer the caller passed — with no error and no warning, so the caller's handle silently holds stale data.

Measured on ttsim v1.10.3 (`libttsim_bh.so`, Blackhole), tt-metal `v0.79.0-dev20260902` debs. Input `7.0`, clamped to `[-1, 1]`, with `out` preallocated and pre-filled with `0.0`:

| bounds | `out` before | `out` after | returned value | returned aliases `out`? |
|---|---|---|---|---|
| tensors (`T(-1)`, `T(1)`) | 0.0 | **0.0** | 1.0 | **False** |
| scalars (`-1.0`, `1.0`) | 0.0 | **1.0** | 1.0 | **True** |

```python
out = ttnn.from_torch(torch.zeros(1,1,32,32), dtype=ttnn.bfloat16, device=d, layout=ttnn.TILE_LAYOUT)
ret = ttnn.clamp(x, min_tensor, max_tensor, output_tensor=out)
# ret[0,0,0,0] == 1.0   correct
# out[0,0,0,0] == 0.0   unchanged — the caller's buffer was never written
```

Same op, same argument, same call shape; only the type of the bounds differs.

## Cause

The two overloads are adjacent in [`unary_composite_op.cpp`](https://github.com/tenstorrent/tt-metal/blob/main/ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_composite_op.cpp).

The scalar overload ([#L185-L216](https://github.com/tenstorrent/tt-metal/blob/main/ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_composite_op.cpp#L185-L216)) names the parameter and forwards it on both return paths:

```cpp
    const std::optional<Tensor>& output_tensor) {
    ...
    return ttnn::clamp_tss(a, min_val, max_val, output_mem_config, output_tensor);
```

The tensor overload ([#L218-L251](https://github.com/tenstorrent/tt-metal/blob/main/ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_composite_op.cpp#L218-L251)) comments the parameter out at the signature and never refers to it again:

```cpp
Tensor clamp(
    const Tensor& input_a,
    std::optional<Tensor> min,
    std::optional<Tensor> max,
    const std::optional<MemoryConfig>& output_mem_config,
    const std::optional<Tensor>& /*output_tensor*/) {
```

All three of its return paths — min-only at [:227](https://github.com/tenstorrent/tt-metal/blob/main/ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_composite_op.cpp#L227), max-only at [:234](https://github.com/tenstorrent/tt-metal/blob/main/ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_composite_op.cpp#L234), both bounds at [:246](https://github.com/tenstorrent/tt-metal/blob/main/ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_composite_op.cpp#L246) — end in a `ttnn::where` whose output argument is left defaulted.

Grepping the identifier in that file returns `:190`, `:211`, `:215` (all inside the scalar overload) and `:223` (the commented-out parameter). There are zero occurrences in the tensor overload's body.

## No fallback

Nothing re-applies the buffer afterwards. The wrapper returns the freshly allocated tensor directly; there is no post-hoc copy, no second pass, and no host-side fixup between the `where` and the return. Nothing validates that the supplied tensor's shape, dtype or memory config match what the op will produce either, since the argument is never read at all.

## Why tests miss it

No test in the repo calls the tensor-bound `ttnn.clamp` together with `output_tensor=`. The tests that do pass `output_tensor` use scalar bounds, which is the overload that works, and the tests that use tensor bounds do not pass an output tensor. Checking the *returned* value passes under the bug — the returned tensor is correct. Only inspecting the caller's buffer afterwards reveals it.

### Every other op I could test honours it

I swept 57 unary and binary ops that accept `output_tensor=` — `abs`, `neg`, `relu`, `exp`, `sigmoid`, `sqrt`, `square`, `tanh`, `gelu`, `silu`, `log`, `reciprocal`, `sign`, `floor`, `ceil`, `round`, `rsqrt`, `erf`, `erfc`, `sin`, `cos`, `atan`, `softplus`, `hardsigmoid`, `hardswish`, `elu`, `celu`, `selu`, `mish`, `digamma`, `lgamma`, `i0`, `expm1`, `log1p`, `log2`, `log10`, `add`, `subtract`, `multiply`, `divide`, `maximum`, `minimum`, `pow`, `squared_difference`, `logical_and`, `logical_or`, `logical_xor`, `bias_gelu`, `atan2`, `hypot`, `xlogy`, `nextafter`, `logaddexp`, `ldexp`, `gcd`, `lcm`, `remainder` — each with a preallocated buffer, checking the buffer afterwards rather than the return value.

**57 honoured it. None ignored it.** Four do not take the argument at all, which is fine — they reject it with a `TypeError` rather than accepting and discarding.

So this is not a general gap in the optional-output plumbing that a caller should expect. It is specific to the tensor-bounds overload of `clamp`, and the scalar-bounds overload of the same op — the natural thing to compare against — gets it right.

## Fix

Name the parameter and forward it. `ttnn::where` already takes an optional output as its final argument and honours it, so all three return paths can pass it through unchanged:

```cpp
-    const std::optional<Tensor>& /*output_tensor*/) {
+    const std::optional<Tensor>& output_tensor) {
...
-    return ttnn::where(..., output_memory_config);
+    return ttnn::where(..., output_memory_config, output_tensor);
```

The middle `temp` computation stays as it is — only the three terminal `where` calls need the argument, since those are what produce the returned value.

If forwarding is not wanted for some paths, the alternative is to reject the combination with a `TT_FATAL` rather than accept the argument and discard it, so a caller relying on in-place semantics finds out.

I would like to work on this through the bounty program, and am happy to prepare the PR.
